README
EPA Region 10 dbf utilities.
Version 5.7 July 13, 1998


This package is a set of utilities for translating data between dBASE
dbf, ASCII, and ARC/INFO formats.  Most of the programs run either as
ARC commands or as system commands independent of ARC/INFO.

An install script is included that configures, builds, tests, and
installs the package.  Pre-compiled packages are available for several
systems.

These programs were developed for EPA use, but have been released as
free software.  Anyone can use them.

A web page is available that describes the package, including changes
since the last version, and links to download the package.  This page
is available from the software section of the EPA Region 10 home page:

  http://www.epa.gov/r10earth/r10.html


Utilities included
------------------

The following utilities are implemented as ARC commands, but they do
not require ARC/INFO, and may also be installed as system commands.

  dbf2info    -- Creates an INFO data file from a dBASE dbf file.
  info2dbf    -- Creates an dBASE dbf file from INFO data file.
  ascii2info  -- Creates an INFO data file from an ASCII file.
  info2ascii  -- Creates an ASCII file from an INFO data file.
  dbflook     -- Lists the header, field defs, and data in a dbf file.
  dbflist     -- Lists the data section of a dbf file to stdout.
  dbfitems    -- Lists the field definitions of a dbf file to stdout.
  infodir     -- Lists the contents of an INFO directory, with wildcarding.


The following are implemented as ARC commands.  They require ARC/INFO.

  info2cover  -- Creates a point coverage from an INFO file.
  dbf2cover   -- Creates a point coverage from a dbf file.


The following utilities are implemented as system commands.  They run
from the system prompt, and do not require ARC/INFO.

  infolook    -- Lists the header, item defs, and data in an INFO file.
  infolist    -- Lists the data section of an INFO data file to stdout.
  infoitems   -- Lists the item definitions of an INFO data file to stdout.
  infodel     -- Deletes INFO files, with wildcarding.


Determining which files you need
--------------------------------

The package is available via anonymous ftp from ftp.epa.gov in
/pub/gistechs/programs.  The files to get are dbf-N.N.readme (this
file) and dbf-N.N.tar.Z (or dbf-N.N.tar.gz if you have gzip), where
"N.N" is the current version number.

The dbf-N.N.tar file is the most generic from of the package.  It has
the complete source code, and everything you need to install the
package on various UNIX boxes.

There may also be tarfiles that contain pre-compiled executables for
specific machines, eg, dbf-N.N.DGUX.tar for Data General AViiON, etc.
These contain no C source code, just the compiled programs.  If there
is a tar for your machine, and you don't want the source, you can save
some time by getting that instead of dbf-N.N.tar.


Loading the package on to your system
-------------------------------------

Create directory where you want to keep the package.  This directory
will remain the permanent home for all the files.  Files will be
accessed from here when the programs are run, thus this directory must
be accessible to everyone whom you want to be able to run the
programs.

If you already have a previous version of the package, choose a
different directory, eg, dbf2info-N.N, where N.N is the current
version.  This will allow you to revert to the previous version if
desired.  Later, you can delete the old version.  If your old version
is 5.0 or higher you might want to run "dbfinstall uninstall" first.
This is not available below version 5.0.

If you received the package as a compressed tarfile, place the file in
the directory you just created, uncompress it and extract the files
from the tar.  If you received the package on tape, extract the tar
into the directory.


Configuring the package for your system
---------------------------------------

If you are installing from a pre-compiled release, skip this section;
it applies only to the full source release.

The install script checks for some system dependencies and configures
the package accordingly, but there are some things it can't check for.

There are two files to check.  The first is rbdpaths.h.  At the top of
the file is a series of variables for various systems.  Not all
systems are represented.  There are instructions in the file to help
you determine which one to use for your system.

The second file to check is PROBLEMS.  This file describes any known
problems that may arise during installation.  If a problem is known to
occur on your system, you will find instructions to fix the problem.


Configuring the install script
------------------------------

There is a script called dbfinstall that handles configuring,
building, testing, and installing the package.  This script is written
to run in the Bourne Shell.

The install script comes set so that all of the programs will be
installed as ARC commands, and none of them will be installed as UNIX
commands.  The last group of programs (infolook, etc) will thus not be
installed at all.

The advantage to doing the UNIX install, is that the programs will
then be available outside of ARC/INFO.  It's quite handy to be able to
list INFO directories, examine INFO files, create and delete them,
without having to start ARC.

If you don't want to do the UNIX install at this time, skip ahead to
the next section.  You can do the UNIX install later if you change
your mind.

However, if you don't want to do the ARC install, you will have to
edit dbfinstall and make the following change:

    change "install_arc=1" to "install_arc=0"

Unless you have chosen not to do the ARC install, the environment
variable ARCHOME must be set and available to the dbfinstall script.
If ARCHOME is not accessible to your shell, you will need to edit
dbfinstall, substituting the full path to ARC's home for ARCHOME.

If you want to do the UNIX install, you will have to edit dbfinstall
and make the following changes:

    change "install_unix=0" to "install_unix=1"

  Figure out where you want the executables to be placed.  This must
  be somewhere on everyone's path, like /usr/global/bin or whatever
  you use at your site.

    then change "binpath=" to the directory you selected.

  If you have an earlier version (5.4 or older) of the utilities
  installed, and you want to uninstall it before installing the new
  version, you will also have to set manpath to whatever you used when
  you installed the earlier version so it can uninstall the man pages.
  At version 5.5, and beyond, man pages are no longer installed.


Running the install script
--------------------------

Type "dbfinstall full".  This will configure the package for your
system, compile everything (unless you're installing from a
pre-compiled release), verify that the programs function correctly,
and install them.

You can do the steps one at a time if you wish.  Type "dbfinstall" to
see the usage for other options.  For more help on using the install
script, type "dbfinstall help".

If you encounter problems, consult the file PROBLEMS, to see if there
is a known fix.  If not, contact me.


Disclaimers
-----------

These programs are not recommended or endorsed by The United States
Government, or by the Environmental Protection Agency.  The programs
are provided "as is" without warranty of any kind, either expressed or
implied.

These programs make extensive use of a public domain C package for
direct INFO access called infolib written by Todd Stellhorn of ESRI.  
The infolib code contains the following disclaimer: 

           This software is being released into the public
           domain.  It may be freely used whole or in part or
           distributed as a whole as long as this notice remains
           attached to the source code.  The author and ESRI
           do not assume any liability or responsibility in the
           use or distribution of this software. 


Randy Deardorff <deardorff.randy@.epa.gov>
US EPA Region 10, Seattle.


CHANGES
Version 5.7 July 13, 1998
---------------------------------------------------------------------

Changes in this version are significant to WinNT/9x only.

Fixed a problem in the InfoLib routine InfoParsePath() to do with
filenames on the PC.  The code was looking for ":" as an arc style
info path, eg, "/foo/info:arc:infofile".  Added some code to recognize
PC style driveletter:/ paths.

Renamed files that had been changed to conform to 8.3.  All files now
have same names in UNIX and NT.
 
Version 5.6 June 9, 1997
---------------------------------------------------------------------
 
Fixed a bug where I and B values greater than 2147483647 were
overflowing and ending up as 2147483647.  Changes in recinfo2dbf and
recinfo2asc.  Effects info2dbf, infolook, infolist, etc.

Fixed a bug in Todd's code similar to above but effecting dbf2info and
others that write info records.  Function is InfoEncode in infoitem.c.

Made info2cover and dbf2cover define the datum of the output coverage
as NAD27, unless a datum is specified in the projection file.

At DGUX 4.11T, the programs and libraries need "gcc -traditional".
This change also requires "-lm".  See Makefile for more.  Modified
makefiles accordingly.  Other systems my need to reverse these
changes, or try -fwritable-strings, or whatever switch your compiler
uses to allow string constants to be written to.

Fixed bug in defsinfo2dbf that caused info2dbf to give a fatal error
for N or F fields with a width of less than 3 and with 0 decimals.

Eliminated all error checking on incoming date fields in ascii2info.
This allows dates to be blank or empty without causing an error.

Version 5.5 April 21, 1995
---------------------------------------------------------------------

Fixed a bug in map/dmsint2dbl.c called indirectly by dbf2cover and
info2cover.  This bug caused bogus results with negative longitudes.

Added record number to error messages in makecoocoo and makegencoo.
These are part of info2cover, and therefore also part of dbf2cover.

Made changes to allow programs to work under DOS for processing
ArcView data.

The programs no longer have ARC help pages or UNIX man pages.  They
now accept a "-help" argument to display a built-in help page.  This
was done so that the help pages would be consistenly available both
ARC 6, and ARC 7, and from the system prompt (UNIX or DOS).  The ARC
help pages and UNIX man pages are no longer supplied or installed.
However, they are still uninstalled in case they were installed by a
previous version.
 
Added a new program, infodel, that deletes INFO files (with
wildcarding) from the system prompt.  If the INFO file is an external
file, eg, cover.pat, only the actual INFO file is deleted, not the
cover/pat file.

Installing the utilities as ARC commands is now optional.  This allows
sites that don't have ARC/INFO to install and use the system level
utilities.

The install script no longer uses ARC/INFO.  Formerly, the
verification tests called ARC to create a workspace and import an INFO
file.  These steps are now done at the system level.  The verification
tests are now part of the "full" install.

Fixed bugs in infodir, mostly to do with optional args.

Fixed a Sun only bug in info2ascii.

Rewrote misc/syscwd.c, which was causing infodir to dump a core on
DEC/Alpha.


Version 5.4 December 16, 1994
---------------------------------------------------------------------
 
Fixed problems having to do with 8.3 workspaces in the info direct
access functions.  The functions now work equally well under ARC 6.1.1
and ARC 7.0.2 regardless of whether the workspace you are in is
pre-8.3 or 8.3, and regardless of whether you are attempting to create
the FIRST info file, or subsequent info files.  This elimates the
workarounds that were required at ARC 7.0.2 in version 5.3.

Fixed bug in dbf2info wherein it attempted to read a directory as if
it were a dbf file.  This fix also applies to dbflook, dbfitems, and
dbflist.

Fixed bugs in infodir.

 
Version 5.3 November 29, 1994
---------------------------------------------------------------------
 
Changed Makefile to refer to the math library as /usr/lib/libm.a.
This should work on more systems than the previous -lm usage.

Fixed three non-DG bugs in my C code.

Help files are not installed unless $ARCHOME/help/arc exists.  It
no longer exists in arc 7.0.2.

This version does NOT fix problems with 8.3 workspaces in arc 7.0.2.
Unfortunately, there are a few problems in the info direct access
functions that I'm using.  See arc 7.0.2 notes in README.


Version 5.2 October 20, 1994
---------------------------------------------------------------------

Incorporated version 1.2 of Todd Stellhorn's info direct access
functions.  This allows the programs to work with the new 8.3 file
name option in ARC 7.0.

 
Version 5.1 October 5, 1994
---------------------------------------------------------------------
 
Added SINGLEQUOTE option to info2ascii and ascii2info.

Added PROBLEMS file with problems and fixes for various systems.

Improved portability of the swaptest program by allowing for variously
sized data types.  Previous version would fail unless long was 4 bytes
and unsigned short was 2 bytes.  The main programs already did this,
but swaptest did not.

Changed a few data types in dbflib for compatibility with PC.


Version 5.0 September 13, 1994
---------------------------------------------------------------------

Incorporated Todd Stellhorn's (tstellhorn@esri.com) info direct access
functions into the supplied infoxlib.a.  See source code in infox for
Todd's notices and disclaimers.

Converted all programs to read and write INFO data files directly
rather than via an E00 file.  Programs are now more than three times
faster when reading INFO files, and over six times faster when writing
them.

Everything except dbf2cover and info2cover can now be run from the

infodir.c
/* infodir.c
   ------------------------------------------------------------------------
   Lists the contents of an INFO directory.
   
   Written: 09/07/94 Randy Deardorff.  USEPA Seattle.
   
   Modification log: (all changes are by the author unless otherwise noted)
   
   12/15/94 -- Updates for 8.3 v pre 8.3 workspaces.  Got rid of
   troublesome cwd call.  Not needed anyway.
   
   12/28/94 Added in-line help.
   
   12/29/94 Fixed bugs in optional arg handling that prevented you
   from listing all files of a certain type, eg, infodir df.  Also
   fixed bug that caused the program to fail if you specified a single
   file, eg infodir cover.pat.  Also fixed bug that caused files that
   did not match filespec to be listed if they matched filetype.  Now
   they must match both.

   1/20/95 Changes to work on DOS.

   02/03/95 Added "%" and "#" alternates for "*" besides "@".

   ------------------------------------------------------------------------ */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <rbdpaths.h>
#include MISC_H
#include INFOLIB_H
#include INFO_H
#include INFODEFS_H

#define TRUE 1
#define FALSE 0
#undef DEBUGGING

int isinfodir();
int help();
int writehelp();

int main(argc, argv)
     int    argc;
     char   *argv[];
{
  short int     numargs, done, curcol, nomatch, match, pos1, pos2, jlen;
  long int      i, pos;
  short int     doallfiles, dowidelist, ignorecase, numitems, reclen;
  short int     typetolist, gousage, newdirtype;
  char          formatstring[80];
  short int     filespecarg = 1, typetolistarg = 2;
  
  char          filespec[128], temp[128], infodirpath[128], filetype[3];
  char          definfopath[128];
  
  char          dirfile[128];
  char          username[USERNAME_SIZE];
  char          filename[INTERNALNAME_SIZE];
  char          tfilename[INTERNALNAME_SIZE];
  char          dirpath[128];
  char          extname[EXTNAME_SIZE+2];
  char          extpath[128];
  char          nitpath[128];
  char          tmppath[128];
  DirRecord     dirrec;
  char          internalname[INTERNALNAME_SIZE];
  FILE          *dirfp = (FILE *) NULL;
  
  char *ftypelist="DF|PG|RP|SF|IF";
  
  
  /* Set program details. */
  
  char *progname="INFODIR";
  char *progargs="{-lc} {files} {DF|PG|RP|SF|IF}";
  char *progver= "5.7 July 13, 1998";
  int  authorid=TRUE;
  
  
  /* Fetch arguments, check validity. */
  
  numargs = argc - 1;

  
  /* Check if help requested. */
  
  if (numargs >= 1)
    {
      strcpy(temp,argv[1]);
      strupr(temp);
      if (strcmp(temp,"-HELP")==0)
        {
          help(progname,FALSE);
          return FALSE;
        }
    }
  
  /* Fetch and set any dash option flags. */
  
  dowidelist=FALSE;
  ignorecase=TRUE;
  gousage=FALSE;
  if (numargs >= 1)
    {
      strcpy(temp,argv[1]);
      if (temp[0] == '-')
        {
          filespecarg++;
          typetolistarg++;
          for (i = 1; i < strlen(temp); i++)
            switch (temp[i])
              {
              case 'l':
                dowidelist=TRUE;
                break;
              case 'c':
                ignorecase=FALSE;
                break;
              default:
                gousage=TRUE;
                break;
              }
        }
    }
  if (gousage)
    {
      usage(progname,progver,progargs,authorid);
      return FALSE;
    }
  
  
  /* Make sure not too many args. */
  
  if (numargs > typetolistarg)
    {
      usage(progname,progver,progargs,authorid);
      return FALSE;
    }


  /* Get the filespec arg. */

  if (numargs >= filespecarg)
    {
      strcpy(filespec,argv[filespecarg]);
      if (strcmp(filespec,"#") == 0 || strcmp(filespec,"@") == 0)
        {
          doallfiles = TRUE;
          filespec[0]='\0';
        }
      else
        {
          doallfiles = FALSE;
          stredt(filespec,'#','*');
          stredt(filespec,'%','*');
          stredt(filespec,'@','*');
          strunquote(filespec);
        }
    }
  else
    {
      doallfiles = TRUE;
      filespec[0]='\0';
      typetolistarg--;
    }


  /* Check if filespec is really a filetype, and adjust accordingly. */

  typetolist = -1;
  strcpy(temp,filespec);
  strupr(temp);
  i=strkeyword(ftypelist,temp,'|',"left");
  if (i >= 1)
    {
      /* Filespec is really a filetype.  */
      
      strcpy(filespec,"*");
      typetolist=i-1;
    }
  else
    {
      /* Filespec is not a file type.  Check for filetype arg. */
      
      if (numargs >= typetolistarg)
        {
          strcpy(temp,argv[typetolistarg]);
          strupr(temp);
          typetolist=strkeyword(ftypelist,temp,'|',"left");
          if (typetolist <= 0)
            {
              usage(progname,progver,progargs,authorid);
              return FALSE;
            }
          else
            {
              doallfiles = FALSE;
              typetolist--;
            }
        }
    }
  
  
  /* Set path to INFO directory, and or cleanup filespec. */
  
  definfopath[0]='\0';
  strcat(definfopath,"info");
  infodirpath[0]='\0';
  if (filespec[0] == '\0')
    {
      /* No path or filespec arg.  Assume current directory. */
      
      strcpy(infodirpath,definfopath);
    }
  else
    {
      if (strfind(filespec,OPSYS_SEP,0) < 0)
        {

#ifdef DEBUGGING
          printf("No separator.  A file or wildcard in current directory.\n");
#endif
          
          /* If filespec has no star, it could be either an info file
             name, or the workspace name.  Have to check. */
          
          if (strfind(filespec,"*",0) < 0)
            {
              
#ifdef DEBUGGING
              printf("No star, could be workspace name.  Checking...\n");
#endif
              syscwd(temp);
              strcat(temp,OPSYS_SEP);
              strcat(temp,filespec);
              strcat(temp,OPSYS_SEP);
              strcat(temp,"info");
              if (isinfodir(temp))
                {
                  strcpy(infodirpath,temp);
                  filespec[0]='\0';
                  doallfiles=TRUE;
                }
              else
                {
                  syscwd(infodirpath);
                  strcat(infodirpath,OPSYS_SEP);
                  strcat(infodirpath,"info");
                }
            }
          else
            {
              
#ifdef DEBUGGING
              printf("Has star, can't be the workspace name.\n");
#endif
              strcpy(infodirpath,definfopath);
            }
        }
      else
        {
          
#ifdef DEBUGGING
          printf("At least one path separator.\n");
#endif
          
          /* Get separate directory and entry name.  Note that we
             remove OPSYS_SEP"info:arc:" from directory. */
          
          sprintf(temp,"%s%s",OPSYS_SEP,"info:arc:");
          pos1 = strfind(filespec,temp,0);
          sprintf(temp,"%s%s",OPSYS_SEP,"info!arc!");
          pos2 = strfind(filespec,temp,0);
          jlen = strlen(filespec);
          if (pos1 >= 0)
            {
              strsubstr(filespec,infodirpath,0,pos1);
              strsubstr(filespec,temp,pos1+10,jlen-pos1-10+1);
              strcpy(filespec,temp);
            }
          else
            if (pos2 >= 0)
              {
                strsubstr(filespec,infodirpath,0,pos2);
                strsubstr(filespec,temp,pos2+10,jlen-pos2-10+1);
                strcpy(filespec,temp);
              }
            else
              {
                strcpy(temp,filespec);
                strentry(temp,filespec);
                strdir(temp,infodirpath);
              }
          
          /* If filespec has no star, it could be either an info file
             name, or the workspace name.  Have to check. */
          
          if (strfind(filespec,"*",0) < 0)
            {
              strcpy(temp,infodirpath);
              strcat(temp,OPSYS_SEP);
              strcat(temp,filespec);
              strcat(temp,OPSYS_SEP);
              strcat(temp,"info");
              if (isinfodir(temp))
                {
                  strcpy(infodirpath,temp);
                  filespec[0]='\0';
                  doallfiles=TRUE;
                }
            }
          
          /* Append OPSYS_SEP"info" if not supplied. */
          
          sprintf(temp,"%s%s",OPSYS_SEP,"info");
          if (strfind(infodirpath,temp,0) < 0 )
            {
              strcat(infodirpath,OPSYS_SEP);
              strcat(infodirpath,"info");
            }
        }
    }
  
  
  /* Ignore case? */
  
  if (ignorecase)
    strlwr(filespec);

  
  /* Args ok.  Continue. */
  
#ifdef DEBUGGING
  printf("infodirpath=%s\n",infodirpath);
  printf("   filespec=%s\n",filespec);
  printf(" dowidelist=%d\n",dowidelist);
  printf(" ignorecase=%d\n",ignorecase);
  printf(" typetolist=%d\n",typetolist);
#endif
  
  
  /* Open the INFO dir file. */
  
#ifdef DEBUGGING
  printf(" infodirpath=%s\n",infodirpath);
#endif
  
  sprintf(dirfile,"%s%s%s%s",infodirpath,OPSYS_SEP,"arc",DIR_EXT);
#ifdef DEBUGGING
  printf("Trying %s\n",dirfile);
#endif
  
  if (sysexist(dirfile))
    {
      if((dirfp = OpenFile(dirfile,InfoREAD)) == (FILE *) NULL)
        {
          printf("%s: Could not open INFO directory file %s.\n",
                 progname,dirfile);
          return FALSE;
        }
      newdirtype=FALSE;
    }
  else
    {
      sprintf(dirfile,"%s%s%s",infodirpath,OPSYS_SEP,"arc.dir");
#ifdef DEBUGGING
      printf("Trying %s\n",dirfile);
#endif

      if (sysexist(dirfile))
        {
          if((dirfp = OpenFile(dirfile,InfoREAD)) == (FILE *) NULL)
            {
              printf("%s: Could not open INFO directory file %s.\n",
                     progname,dirfile);
              return FALSE;
            }
          newdirtype=TRUE;
        }
      else
        {
          printf("%s: Could not open INFO directory file.\n",progname);
          return FALSE;
        }
    }
  
#ifdef DEBUGGING
  printf("OpenFile ok.\n");
#endif
  
  
  /* Set up. */
  
  pos=0;
  done=FALSE;
  curcol=1;
  nomatch=TRUE;
  if (doallfiles)
    match=TRUE;
  else
    match=FALSE;
  if (dowidelist)
    {
      printf(" Type Name                            Internal Name     Records");
      printf(" Length External\n");
      if (sizeof(long)==(sizeof(dirrec.NumberRecords)))
        if (newdirtype)
          strcpy(formatstring,"  %s  %-32s  %-9s %12ld %6d    %c%c\n");
        else
          strcpy(formatstring,"  %s  %-32s  %-9s %13ld %6d    %c%c\n");
      else
        if (newdirtype)
          strcpy(formatstring,"  %s  %-32s  %-9s %12d %6d    %c%c\n");
        else
          strcpy(formatstring,"  %s  %-32s  %-9s %13d %6d    %c%c\n");
    }
  
  
  /* List files until eof or error. */
  
  while(!done)
    if(!DirReadRecord(dirfp,&dirrec,pos))
      done=TRUE;
    else
      {
        if (!dirrec.EmptyFlag)
          {

            /* Get internal (user level) file name. */
            
            strsubstr(dirrec.InternalName,filename,0,INTERNALNAME_SIZE-1);
            strtrim(filename,"both",' ');

            
            /* Check if this file matches our search criteria. */

            if (!doallfiles)
              if (ignorecase)
                {
                  strcpy(temp,filename);
                  strlwr(temp);
                  match=strmatch(temp,filespec);
                }
              else
                match=strmatch(filename,filespec);
            
            if (match)
              if (typetolist >= 0)
                match=(typetolist==dirrec.FileType);
            
            
            /* Print the filename to screen. */
            
            if (match)
              {
                if (dowidelist)
                  
                  /* Single column listing pretty much like INFO does. */
                  
                  {
                    
                    /* Get external name.  We have to add the three letter
                       tag for DAT or SRC.  Type DF files are DAT.  All others
                       are SRC. */
                    
                    strsubstr(dirrec.ExternalName,extname,0,EXTNAME_SIZE-1);
                    strtrim(extname,"both",' ');
                    if (dirrec.FileType == 0)
                      strcat(extname,"DAT");
                    else
                      strcat(extname,"SRC");
                    
                    /* Get the file type from our list. */
                    
                    strext(ftypelist,dirrec.FileType+1,filetype,'|');
                    
                    /* Print information for this file. */
                    
                    printf(formatstring,
                           filetype,
                           filename,
                           extname,
                           dirrec.NumberRecords,
                           dirrec.RecordLength,
                           dirrec.ExternalFlag[0],
                           dirrec.ExternalFlag[1]);
                    
                  }
                else
                  
                  /* Two column listing. */
                  
                  {
                    if (curcol == 1)
                      {
                        printf("%-32s",filename);
                        curcol++;
                      }
                    else
                      {
                        printf(" %-32s\n",filename);
                        curcol=1;
                      }
                  }
                nomatch=FALSE;
              }
          }
        pos++;
      }
  
  
  /* Wrap up. */
  
  if (!dowidelist && curcol == 2)
    putchar('\n');

  if (nomatch)
    if (doallfiles)
      printf("No files.\n");
    else
      printf("No match.\n");
  
  
  /* Close the INFO directory file. */
  
  if(!CloseFile(dirfp))
    {
      printf("%s: Could not close INFO directory file: %s.\n",progname,dirfile);
      return FALSE;
    }
  
  return TRUE;
}


int isinfodir(infodirpath)
     char *infodirpath;
{
  char          infodir[128];
  char          dirfile[128];
  char          username[USERNAME_SIZE];
  char          filename[INTERNALNAME_SIZE];
  FILE          *dirfp = (FILE *) NULL;
  short int     retval;
  
#ifdef DEBUGGING
  printf("  Checking if %s is an info directory...\n",infodirpath);
#endif
  
  sprintf(dirfile,"%s%s%s%s",infodirpath,OPSYS_SEP,"arc",DIR_EXT);

#ifdef DEBUGGING
  printf("    Trying %s...\n",dirfile);
#endif

  if (sysexist(dirfile))
    {
      if((dirfp = OpenFile(dirfile,InfoREAD)) != (FILE *) NULL)
        {
          CloseFile(dirfp);
          retval = TRUE;
        }
      else
        retval = FALSE;
    }
  else
    {
      sprintf(dirfile,"%s%s%s",infodirpath,OPSYS_SEP,"arc.dir");

#ifdef DEBUGGING
      printf("    Trying %s...\n",dirfile);
#endif

      if (sysexist(dirfile))
        {
          if((dirfp = OpenFile(dirfile,InfoREAD)) != (FILE *) NULL)
            {
              CloseFile(dirfp);
              retval = TRUE;
            }
          else
            retval = FALSE;
        }
      else
        retval = FALSE;
    }
  
#ifdef DEBUGGING
  if (retval)
    printf("    Yes, %s opened ok.\n",dirfile);
  else
    printf("    No.\n");
#endif
  
  return retval;
}


/* Help function. */

int help (progname)
     char       *progname;
{
  
  FILE *io;
  char ftag[12], filename[64], cmd[128];
  
  
  /* Set filename. */
  
  if (!sysdate("ftag",ftag))
    {
      printf("%s: Could not generate temporary file name.\n",progname);
      return FALSE;
    }
  sprintf(filename,"xx%s",ftag);
  strlwr(filename);
  
  
  /* Open the file for writes. */
  
  if ((io = fopen(filename, "w")) == NULL)
    {
      printf("%s: Could open %s for writing.\n",progname,filename);
      return FALSE;
    }
  
  
  /* Write the help text to the file. */
  
  writehelp(io);  
  
  
  /* Close the file. */
  
  fclose(io);
  
  
  /* Display it, then delete it. */

#if ANY_UNIX
  sprintf(cmd,"more %s",filename);
#else
  sprintf(cmd,"page %s",filename);
#endif  
  system(cmd);
  sysdel(filename);
  
  
  /* Done. */
  
  return 1;
}

int writehelp (io)
     FILE *io;
{
  /* @ begin help */
  fprintf(io,"Usage: infodir {-lc} {files} {DF|PG|RP|SF|IF}\n");
  fprintf(io,"\n");
  fprintf(io,"Lists the contents of an INFO directory without starting an INFO\n");
  fprintf(io,"session.  Runs from UNIX or DOS level.  Can also be run from ARC via\n");
  fprintf(io,"&system.\n");
  fprintf(io,"\n");
  fprintf(io,"\n");
  fprintf(io,"Arguments:\n");
  fprintf(io,"\n");
  fprintf(io,"{-lc} -l Long listing.  Default is shorter two column list.\n");
  fprintf(io,"      -c Consider case.  Default is ignore case.\n");
  fprintf(io,"\n");
  fprintf(io,"{files} -- Files to list.  See notes.\n");
  fprintf(io,"\n");
  fprintf(io,"{DF|PG|RP|SF|IF} -- Type of INFO files to list.  Default is all types.\n");
  fprintf(io,"\n");
  fprintf(io,"\n");
  fprintf(io,"Notes:\n");
  fprintf(io,"\n");
  fprintf(io,"If no arguments are given, all INFO files in the current workspace's\n");
  fprintf(io,"INFO directory are listed in two column format.\n");
  fprintf(io,"\n");
  fprintf(io,"You can specify files to list either by full name, or by partial name\n");
  fprintf(io,"with wildcard characters.  On UNIX, the customary wildcard character,\n");
  fprintf(io,"'*', works only if you enclose the file specification in single\n");
  fprintf(io,"quotes.  For example:\n");
  fprintf(io,"\n");
  fprintf(io,"    infodir '*.pat'\n");
  fprintf(io,"\n");
  fprintf(io,"If you prefer, you can substitute '@' or '%%' for the wildcard.  In\n");
  fprintf(io,"this case, you don't need to quote the file specification.\n");
  fprintf(io,"\n");
  fprintf(io,"You can specify another directory to list, giving either a full path,\n");
  fprintf(io,"or a relative path.  For example:\n");
  fprintf(io,"\n");
  fprintf(io,"    infodir /usr/workspace/cover.pat\n");
  fprintf(io,"    infodir workspace/cover.pat\n");
  fprintf(io,"\n");
  fprintf(io,"You can use /info:arc: if you want, but you don't need to.  For example,\n");
  fprintf(io,"the following are equivalent:\n");
  fprintf(io,"\n");
  fprintf(io,"    infodir /usr/workspace/info:arc:cover.pat\n");
  fprintf(io,"    infodir /usr/workspace/cover.pat\n");
  fprintf(io,"\n");
  fprintf(io,"See also infodel, infolist, infoitems, and infolook.\n");
  /* @ end help */
  return 1;
}
